### DISCUSSION TITLE
Enable user supplied id for connection init, connection close and create subscription

### DISCUSSION THREAD

**Discussion by gowthambala** [graphql-dotnet/graphql-client#587] [Q&A]
Hello Everyone,

 Firstly thank you for creating this awesome library!. I am using the graphql client to talk to a third party graphql server implementation. I am using the client only for subscriptions and not for queries or mutations. The third party server, unfortunately, requires an unique ID to be sent with connection init, subscribe and complete messages.  something like below
 
{_**"id":"bd60c1937b054b288731820ae7f980a8"**_,type":"connection_init","payload":{""....

{_**"id":"bd60c1937b054b288731820ae7f980a8"**_,"type":"subscribe","payload":{""....

{_**"id":"bd60c1937b054b288731820ae7f980a8"**_,"type":"complete"}

but currently we send an id only with create subscription. but that too, the id seems to be autogenerated.

var startRequest = new GraphQLWebSocketRequest
            {
                Id = Guid.NewGuid().ToString("N"),
                Type = GraphQLWebSocketMessageType.GQL_SUBSCRIBE,
                Payload = preprocessedRequest
            };

Is it possible to manually send in this ID ? and is there a way I can get this ID to go out as part of all the three messages? connection_init,subscribe and complete?

**Comment by rose-a** [comment]
Hi, the ID is already part of each request sent to the server. Corresponding `subscribe` and `complete` are already using the same id (makes sense, otherwise the server won't know which subscription is completed/stopped).

The *ID* in the protocol refers to the "operation", not to the websocket connection. Therefore `connection_init` or different subscriptions on a single websocket connection all have their own unique id.

It is not possible to set this ID externally.

**Reply by gowthambala** [reply]
Thank you for your reply! Any idea how I can see this id? I am using fiddler to capture the live traffic. I only see the id for subscription request sent and not for connection init or connection end? Am I missing something obvious here? 

{"type":"connection_init","payload":{ --- connection init
{"id":"aebfc4a9594c4c40b8093e21ee45581b","type":"subscribe","payload":{ ---subscribe
{"type":"complete"} - connection complete

![image](https://github.com/graphql-dotnet/graphql-client/assets/7053291/169fcd33-d24a-435a-b232-87ffcb0e5679)

**Reply by rose-a** [reply]
Ok, well I have to correct myself... 

IDs are only created and sent in GraphQL request or subscription contexts (to allow correlation of server responses etc.). If you unsubscribe from a active subscription, the complete request will contain the corresponding subscription ID (see [here](https://github.com/graphql-dotnet/graphql-client/blob/344a3e9f2ff2d110d331d377c25611e0306ed042/src/GraphQL.Client/Websocket/GraphQLWSProtocolHandler.cs#L44), to only complete this exact subscription, not others which might be running on the same websocket connection).

If you dispose of the whole websocket connection, then no ID is sent with complete (although I'm not sure that this is even specified).

The `graphql-transport-ws` implementation in this library aims to stay as close as possible to the [specification](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md), which also uses the ID only to correlate requests and responses:

> `id` used for uniquely identifying server responses and connecting them with the client's requests

If you want to change this behavior, you need to modify this library to create and use your own `GraphQLTransportWSProtocolHandler` [here](https://github.com/graphql-dotnet/graphql-client/blob/344a3e9f2ff2d110d331d377c25611e0306ed042/src/GraphQL.Client/Websocket/GraphQLHttpWebSocket.cs#L375)

**Reply by gowthambala** [reply]
Thank you for your reply ! That is what I ended up doing in order to pass the ID.
